Add DAI support message routing example.

michaelwsherman opened the pull request:
Email support routing solution, built for DAI service kit. Technically customer complaint resolution prediction, but use cases are very, very similar. Built using a public dataset.

Uses BigQuery for the data pipeline and trains an AutoML Tables model. Data pipeline is config+SQL based, making modifications to new data sources manageable. 

Engineering Council bug ID: b/141161346

sahanasub commented:
@googlebot I consent

mos397 commented:
@googlebot I consent

michaelwsherman commented:
Ready for review.

michaelwsherman commented:
Ok, for real ready for review.

TheMichaelHu commented on examples/cloudml-support-routing/scripts/create_features.py:
```diff
@@ -0,0 +1,123 @@
+  # Create the dataset to hold data for the pipeline run.
+  utils.create_dataset(
+      destination_project=global_config['destination_project_id'],
+      destination_dataset=global_config['destination_dataset']
```
nit: add trailing commas per [go/pystyle#trailing-comma](go/pystyle#trailing-comma)

michaelwsherman replied:
fixed a bunch of these

TheMichaelHu commented on examples/cloudml-support-routing/scripts/predict.py:
```diff
@@ -0,0 +1,129 @@
+    logging.warning("%d rows in the batch prediction job failed.",
+                    failed_predictions_table.num_rows)
+
+  # Delete the dataset created by AutoML, catches Forbidden exception raised if
```
Cleaning up tables-made datasets is awesome!

michaelwsherman replied:
:)

TheMichaelHu commented on examples/cloudml-support-routing/scripts/predict.py:
```diff
@@ -0,0 +1,129 @@
+
+
+def main():
+  """Executes batch prediction using a model trained on AutoML Forecasting.
```
This example uses AutoML Tables Forecasting? Or is this a typo and its meant to say AutoML Tables?

michaelwsherman replied:
Good catch. Found a few of these and fixed them.

TheMichaelHu commented on examples/cloudml-support-routing/config/pipeline.yaml:
```diff
@@ -0,0 +1,208 @@
+# Copyright 2020 Google LLC
```
fyi, optional: one pattern I've seen is creating a `pipeline.yaml.example` file with generic contents like this and gitignoring the `pipeline.yaml` file. This way devs can maintain local configs without git yelling at them.

michaelwsherman replied:
This is a good idea, but since you've said optional I'm going to ignore because of the readme changes required. We do say in the readme to copy the file and recommend my_config.yaml as the name, but what you propose is def better.

TheMichaelHu commented on examples/cloudml-support-routing/README.md:
```diff
@@ -0,0 +1,133 @@
+1. Deploy the model: `bash online_predict.sh config/my_config.yaml deploy`. Take note of the "name" value in the response.
+1. Deployment will take up to 15 minutes. To check the status of the deployment run the following command, replacing "operation-name" with the "name" value from the previous step: `curl -X GET -H "Authorization: Bearer $(gcloud auth application-default print-access-token)" -H "Content-Type: application/json" https://automl.googleapis.com/v1beta1/operation-name`. When the operation is complete, th ...
+1. Make a prediction using the provided sample predict_payload.json, containing features of an unlabeled example complaint: `bash online_predict.sh config/my_config.yaml predict predict_payload.json`. The response will have the different clasess with values based on the confidence of the class. To predict for different features, change "values" in the .json file. The order of features in the json ...
+1. You should undeploy your model when finished to avoid excessive charges. Run: `bash online_predict.sh config/my_config.yaml undeploy`.
```
optional: may be good to tell users to check their model in the UI just to make sure they will no longer be charged.

michaelwsherman replied:
good suggestion. done!

michaelwsherman commented on examples/cloudml-support-routing/queries/train_and_predict_split.sql:
```diff
@@ -0,0 +1,25 @@
+  ToSplit.complaint_id,
+  CASE
+    WHEN
+        RAND() < 1-{test_threshold}
```
good catch. fixed.

michaelwsherman commented on examples/cloudml-support-routing/queries/training_features.sql:
```diff
@@ -0,0 +1,61 @@
+-- =============================================================================
+/** Query to create the training features table and the manual train-validation split.*/
+
+WITH GroupSize AS
```
fixed, good catch

michaelwsherman commented on examples/cloudml-support-routing/queries/training_features.sql:
```diff
@@ -0,0 +1,61 @@
+    SELECT
+      company_response_to_consumer
+    FROM  
+      `{destination_project_id}.{destination_dataset}.{clean_table}`
```
I've made a formatting pass of the sql files, fixed a handful of things, should be better now.

michaelwsherman commented on examples/cloudml-support-routing/queries/training_features.sql:
```diff
@@ -0,0 +1,61 @@
+  Training AS
+    (
+      SELECT
+        TrainingFeatures.*,
```
fixed

michaelwsherman commented on examples/cloudml-support-routing/queries/training_features.sql:
```diff
@@ -0,0 +1,61 @@
+-- =============================================================================
+/** Query to create the training features table and the manual train-validation split.*/
+
+WITH GroupSize AS
```
good suggestion. fixed

michaelwsherman commented on examples/cloudml-support-routing/queries/remove_nulls.sql:
```diff
@@ -0,0 +1,28 @@
+FROM 
+  `{source_project_id}.{source_dataset}.{source_table}`
+WHERE 
+  `{}` IS NOT NULL
```
this is a pretty annoying thing to fix, so since you've made it a nit i'm going to ignore. there's actually a list generated that goes here.

michaelwsherman commented:
@TheMichaelHu I've addressed all your review comments. I can't approve the PR until you approve the review, but once you click the buttons we are good to merge. Thank you!
